components/scream/src/physics/p3/tests/p3_subgrid_variance_scaling_unit_tests.cpp

```diff
@@ -0,0 +1,221 @@
+    
+    // Sync relvars to device
+    view_1d<Spack> relvars_device("relvars",1);
+    auto relvars_host = Kokkos::create_mirror_view(relvars_device);    
+    for (Int s = 0; s < Spack::n; ++s) {
+      relvars_host(0)[s] = relvars_info[s];}
+
+    Kokkos::deep_copy(relvars_device, relvars_host);
+    
+    //Get view of output from C++ fn call. 
+    view_1d<Spack> scaling_device("c_scaling", 1);
+    auto scaling_host = Kokkos::create_mirror_view(scaling_device);
+    //note no copy here - just need to get output *back* from device.
+
+    Spack f_scaling; //function output from F90 on CPU
+    //input data for F90 needs to be a struct defined in p3_functions_f90.hpp
+    SubgridVarianceScalingData f_data;
+    
+    //Loop over exponent values and make sure BFB is maintained in each case.
+    Real expons[3] = {1.0,2.47,0.1};
+    for (Int i = 0; i < 3; ++i) {
+
+      // Call Fortran version
+      for (Int j = 0; j < Spack::n; ++j) {
+	f_data.relvar=relvars_host(0)[j];
+	f_data.expon =expons[i];
+        f_scaling[j] = subgrid_variance_scaling(f_data);
+      }
+
+      // Get expon value onto device
+      view_1d<Scalar> expon_device("expon_device", 1);
+      auto expon_host = Kokkos::create_mirror_view(expon_device);
+      expon_host(0)=expons[i]; //this line causes build errors
+      Kokkos::deep_copy(expon_device, expon_host);
+      
+      // Run the lookup from a kernel and copy results back to host
+      Kokkos::parallel_for(RangePolicy(0, 1), KOKKOS_LAMBDA(const Int& j) {
+	  
+	  // Call the function on C++
+	  scaling_device(0) = Functions::subgrid_variance_scaling(
+		              relvars_device(0),expon_device(0) );
+	  
+	}); //end of parallel for
+      
+      // Copy results back to host
+      Kokkos::deep_copy(scaling_host, scaling_device);
+      
+      // Validate results
+      //Scalar c_scaling;
+      for (Int s = 0; s < Spack::n; ++s) {
+
+	REQUIRE(f_scaling[s] == scaling_host(0)[s] );
+      }
+    } //end loop over expons[i]
+
+  } //end function run_bfb_tests
+
+  //-----------------------------------------------------------------
+  KOKKOS_FUNCTION static void subgrid_variance_scaling_linearity_test(const Scalar& relvar,
+    int& errors){
+    //If expon=1, subgrid_variance_scaling should be 1 
+
+    Scalar tol = (util::is_single_precision<Scalar>::value ) ? C::Tol*2 : C::Tol;
+    
+    //Get value from C++ code
+    const Spack relvars(relvar);
+    Spack c_scaling = Functions::subgrid_variance_scaling(relvars,1.0);
+    
+    if ( std::abs(c_scaling[0] -  1) > tol ){
+      printf("subgrid_variance_scaling should be 1 for expon=1, but is %e\n",c_scaling[0]);
+	errors++;}
+  }
+  
+  //-----------------------------------------------------------------
+  KOKKOS_FUNCTION static void subgrid_variance_scaling_relvar1_test(int& errors){
+    //If relvar=1, subgrid_variance_scaling should be factorial(expon)  
+    
+    Scalar tol = (util::is_single_precision<Scalar>::value ) ? C::Tol*2 : C::Tol;
```

Comment:
Arguably, our tolerances should be traits associated with our Scalar type and not just expressions of a single tolerance in physics constants. Something like `utils::precision<Scalar>::Tolerance`. This isn't specific to this PR, though, and it's just an idea that probably needs more exploration.

Reply:
@jeff-cohere - I think we already do that via util::is_single_precision<Scalar>, right? This gives machine epsilon for double or single precision floating point numbers depending on what the type of Scalar is...

Reply:
I agree with the approach of machine precision based on the type. What I meant was that we could implement this same thing by having tolerances for single- and double-precision without having to articulate the Tol^2 and Tol stuff everywhere we use tolerances.

Reply:
@PeterCaldwell `is_single_precision` only tells you whether the build is single or double precision. This line is indeed questionable, since it says that single precision tolerance is twice the one we use in double precision. Either the DP one is "coarse", or the single one is too restrictive (or both). I don't know what kind of precision we expect from these tests, but I assume (hope?) that using double precision would do better than just halving the error. 

@jeff-cohere `Precision<Scalar>` seems like a good utility candidate for ekat.